lib/fs.js

```diff
@@ -62,7 +64,72 @@ if (version[0] < 10 || (version[0] === 10 && version[1] <= 11)) {
   };
 }
 
+// Recursively remove directory
+//   file <string> file or directory to be removed
+//   cb <Function> callback
+const rmRecursive = (file, cb) => {
+  fs.stat(file, (err, stats) => {
+    if (err) {
+      cb(err);
+      return;
+    }
+    if (stats.isDirectory()) {
+      fs.readdir(file, (err, files) => {
+        if (err) {
+          cb(err);
+          return;
+        }
+        const rm = files.reduce(
+          (callback, f) => {
+            return err => {
+              if (err) cb(err);
+              else rmRecursive(path.join(file, f), callback);
+            };
+          },
+          err => {
+            if (err) cb(err);
+            else fs.rmdir(file, cb);
+          }
+        );
+        rm();
+      });
+    } else {
+      fs.unlink(file, cb);
+    }
+  });
+};
+
+// TODO(SemenchenkoVitaliy): remove and use `fs.promises` instead when
+// Node.js 8 is dropped
+let fsPromises;
+if (process.version.slice(1).split('.')[0] >= 10) {
+  fsPromises = fs.promises;
+} else {
+  const list = ['stat', 'readdir', 'rmdir', 'unlink'];
+  fsPromises = iter(list).collectWith(
+    {},
+    (obj, name) => (obj[name] = util.promisify(fs[name]))
+  );
+}
+
+// Recursively remove directory
+//   file <string> file or directory to be removed
+const rmRecursivePromise = async file => {
+  const stats = await fsPromises.stat(file);
```

Comment:
Is there actually a need to `fs.stat()` the file/directory?
Can't we just use `fs.readdir()`, then check if the error is `ENOTDIR` and perform `fs.unlink()` in that case?

Reply:
IMO, it will look less readable. Compare
```js
  const stats = await fsPromises.stat(file);
  if (stats.isDirectory()) {
    const files = await fsPromises.readdir(file);
    await Promise.all(files.map(f => rmRecursivePromise(path.join(file, f))));
    return fsPromises.rmdir(file);
  } else {
    return fsPromises.unlink(file);
  }
```
and

```js
  let files;
  try {
    files = await fsPromises.readdir(file);
  } catch (err) {
    if (err.code === 'ENOTDIR') {
      return fsPromises.unlink(file);
    } else {
      throw err;
    }
  }
  await Promise.all(files.map(f => rmRecursivePromise(path.join(file, f))));
  return fsPromises.rmdir(file);
```

Reply:
You can write it the following way:
```js
const rmRecursivePromise = async file =>
  fsPromises.readdir(file).then(async files => {
    await Promise.all(files.map(f => rmRecursivePromise(path.join(file, f))));
    return fsPromises.rmdir(file);
  }, err => {
    if (err.code === 'ENOTDIR') return fsPromises.unlink(file);
    throw err;
  }); 
```

This is both shorter and, in my opinion, easier to understand, than using try/catch.

Reply:
BTW, shouldn't we also support Windows, since this code will be used in [@metarhia/filestorage](https://github.com/metarhia/filestorage) and on Windows error code will be `ENOENT`?

Reply:
@SemenchenkoVitaliy, yes, sure.